File: src/score/consistency.ts
/**
 * CG02 -- character-consistency scoring.
 *
 * Computes a per-shot visual embedding and flags cross-shot pairs (claiming
 * to be the same named character) whose cosine similarity falls below a
 * threshold derived from this repo's own fixture run (see
 * src/score/testdata/ and CHANGELOG.md -- never an invented number).
 *
 * IMPORTANT, stated plainly and repeated in the CLI output, the JSON
 * `reason` field, and the README: this technique is best-validated on
 * photorealistic/live-action content. Its accuracy on stylized or
 * anime-adjacent character designs -- the dominant visual style in the
 * short-drama category this tool targets -- is UNVERIFIED. Treat a flag on
 * stylized footage as "worth a second look," never as a confirmed defect.
 *
 * Runtime note: the original design called for `onnxruntime-node` +
 * a pretrained ONNX embedding model, flagging a known risk that a prior,
 * unrelated prototype in a prior, unrelated prototype hit a native-binding crash
 * with `@tensorflow/tfjs-node` on Node 24 / macOS ARM. `onnxruntime-node`
 * was smoke-tested directly on that same machine/Node combination before
 * writing this file (an InferenceSession loaded and ran real inference
 * successfully) and did NOT reproduce that failure, so v0.1 uses
 * `onnxruntime-node` as originally planned -- no WASM fallback was needed
 * for the runtime. The one real substitution made is the *model*: see
 * src/score/models/NOTICE.md for why MobileNetV2 (a generic ImageNet
 * feature extractor) is used instead of a dedicated face/CLIP embedding
 * model, and what that means for accuracy.
 */

import { fileURLToPath } from 'node:url';
import { dirname, join } from 'node:path';
import * as ort from 'onnxruntime-node';
import { FRAME_SIZE, readRawFrame } from '../ingest/ffmpeg.js';

const __dirname = dirname(fileURLToPath(import.meta.url));
const MODEL_PATH = join(__dirname, 'models', 'mobilenetv2-7.onnx');

// ImageNet normalization constants used by the bundled MobileNetV2 model's
// documented preprocessing (see src/score/models/NOTICE.md).
const IMAGENET_MEAN = [0.485, 0.456, 0.406];
const IMAGENET_STD = [0.229, 0.224, 0.225];

/**
 * Cross-shot cosine-similarity threshold below which a shot is flagged as
 * a likely character-consistency break. Derived from this repo's own
 * fixture run -- see CHANGELOG.md "CG02 fixture calibration" entry for the
 * exact command and raw numbers this value came from. Not the execution
 * plan's illustrative 0.62 placeholder.
 */
export const CONSISTENCY_SIMILARITY_THRESHOLD = 0.88;

let cachedSession: ort.InferenceSession | undefined;

/** Loads (and caches) the bundled ONNX embedding session. Never fetches
 * anything over the network -- the model file ships inside this package. */
export async function getEmbeddingSession(): Promise<ort.InferenceSession> {
  if (!cachedSession) {
    cachedSession = await ort.InferenceSession.create(MODEL_PATH);
  }
  return cachedSession;
}

/** Converts a raw HWC RGB24 frame buffer into the CHW, ImageNet-normalized
 * Float32Array MobileNetV2 expects. */
export function preprocessFrame(raw: Uint8Array): Float32Array {
  const size = FRAME_SIZE;
  const out = new Float32Array(3 * size * size);
  const planeSize = size * size;
  for (let pixel = 0; pixel < planeSize; pixel++) {
    const r = raw[pixel * 3] as number;
    const g = raw[pixel * 3 + 1] as number;
    const b = raw[pixel * 3 + 2] as number;
    out[pixel] = (r / 255 - IMAGENET_MEAN[0]!) / IMAGENET_STD[0]!;
    out[planeSize + pixel] = (g / 255 - IMAGENET_MEAN[1]!) / IMAGENET_STD[1]!;
    out[2 * planeSize + pixel] = (b / 255 - IMAGENET_MEAN[2]!) / IMAGENET_STD[2]!;
  }
  return out;
}

/** Runs one frame through the embedding model and returns its raw 1000-d
 * output vector (see src/score/models/NOTICE.md for why this is the raw
 * classifier-logit vector rather than a dedicated embedding-layer output). */
export async function embedFrame(
  raw: Uint8Array,
  session: ort.InferenceSession
): Promise<Float32Array> {
  const input = preprocessFrame(raw);
  const tensor = new ort.Tensor('float32', input, [1, 3, FRAME_SIZE, FRAME_SIZE]);
  const inputName = session.inputNames[0];
  if (!inputName) {
    throw new Error('Embedding model exposes no input tensor names');
  }
  const results = await session.run({ [inputName]: tensor });
  const outputName = session.outputNames[0];
  if (!outputName) {
    throw new Error('Embedding model exposes no output tensor names');
  }
  const output = results[outputName];
  if (!output) {
    throw new Error('Embedding model produced no output tensor');
  }
  return Float32Array.from(output.data as Float32Array);
}

/** Computes one embedding per shot by averaging the embeddings of every
 * sampled frame in that shot -- smooths single-frame noise into a more
 * stable per-shot signal than picking any single frame would. */
export async function embedShot(
  framePaths: string[],
  session: ort.InferenceSession
): Promise<Float32Array> {
  if (framePaths.length === 0) {
    throw new Error('Cannot embed a shot with zero extracted frames');
  }
  let sum: Float32Array | undefined;
  for (const framePath of framePaths) {
    const raw = await readRawFrame(framePath);
    const embedding = await embedFrame(raw, session);
    if (!sum) {
      sum = new Float32Array(embedding.length);
    }
    for (let i = 0; i < embedding.length; i++) {
      sum[i] = (sum[i] as number) + (embedding[i] as number);
    }
  }
  const result = sum as Float32Array;
  for (let i = 0; i < result.length; i++) {
    result[i] = (result[i] as number) / framePaths.length;
  }
  return result;
}

export function cosineSimilarity(a: Float32Array, b: Float32Array): number {
  if (a.length !== b.length) {
    throw new Error('Cannot compare embeddings of different dimensionality');
  }
  let dot = 0;
  let normA = 0;
  let normB = 0;
  for (let i = 0; i < a.length; i++) {
    const av = a[i] as number;
    const bv = b[i] as number;
    dot += av * bv;
    normA += av * av;
    normB += bv * bv;
  }
  if (normA === 0 || normB === 0) {
    return 0;
  }
  return dot / (Math.sqrt(normA) * Math.sqrt(normB));
}

/**
 * Infers the named character a clip belongs to from its filename, using
 * a `<character>_<shot-id>.<ext>` convention (e.g. "mei_shot01.mp4" ->
 * "mei"). This is a deliberate v0.1 simplification: there is no
 * industry-standard character-tagging metadata format across AI
 * short-drama pipelines, so ContinuityGuard reads it from the filename
 * rather than requiring a separate manifest. Clips that don't match the
 * convention are grouped under a single "unlabeled" bucket and are never
 * compared to each other for consistency (no named-character claim to
 * check them against).
 */
export function parseCharacterName(clipFileName: string): string | undefined {
  const withoutExt = clipFileName.replace(/\.[^.]+$/, '');
  const match = /^([a-zA-Z][a-zA-Z0-9]*)_/.exec(withoutExt);
  return match?.[1]?.toLowerCase();
}

export interface ShotInput {
  clip: string;
  framePaths: string[];
}

export interface ConsistencyFlag {
  clip: string;
  character: string;
  referenceClip: string;
  similarityScore: number;
  reason: string;
}

export interface ConsistencyResult {
  charactersTracked: number;
  flaggedShots: ConsistencyFlag[];
}

/** Builds the honest, unverified-content-labeled reason string for a
 * flagged shot -- kept as its own function so every call site (CLI, JSON
 * report) uses identical, never-drifting wording. */
export function buildConsistencyReason(similarity: number): string {
  return (
    `Cross-shot embedding similarity ${similarity.toFixed(2)} is below the ` +
    `${CONSISTENCY_SIMILARITY_THRESHOLD.toFixed(2)} threshold. Best-validated on ` +
    'photorealistic content; accuracy on stylized/anime-adjacent character designs ' +
    'is unverified -- treat this as a prompt for human review, not a confirmed defect.'
  );
}

/**
 * Pure grouping + flagging logic, independent of the embedding model --
 * given already-computed per-shot embeddings, groups by character and
 * flags any shot whose similarity to that character's first-seen
 * ("reference") shot falls below CONSISTENCY_SIMILARITY_THRESHOLD.
 * Split out from `scoreConsistency` so this logic has fast, deterministic
 * unit-test coverage without depending on real ONNX inference for every
 * test case.
 */
export function flagInconsistentShots(
  shots: { clip: string; embedding: Float32Array }[],
  threshold: number = CONSISTENCY_SIMILARITY_THRESHOLD
): ConsistencyResult {
  const byCharacter = new Map<string, { clip: string; embedding: Float32Array }[]>();
  for (const shot of shots) {
    const character = parseCharacterName(shot.clip);
    if (!character) continue;
    const bucket = byCharacter.get(character) ?? [];
    bucket.push(shot);
    byCharacter.set(character, bucket);
  }

  const flaggedShots: ConsistencyFlag[] = [];
  for (const [character, bucket] of byCharacter) {
    if (bucket.length < 2) continue;
    const reference = bucket[0] as { clip: string; embedding: Float32Array };
    for (let i = 1; i < bucket.length; i++) {
      const candidate = bucket[i] as { clip: string; embedding: Float32Array };
      const similarity = cosineSimilarity(reference.embedding, candidate.embedding);
      if (similarity < threshold) {
        flaggedShots.push({
          clip: candidate.clip,
          character,
          referenceClip: reference.clip,
          similarityScore: Number(similarity.toFixed(4)),
          reason: buildConsistencyReason(similarity),
        });
      }
    }
  }

  return {
    charactersTracked: byCharacter.size,
    flaggedShots,
  };
}

/** End-to-end CG02 entry point: computes embeddings for every shot (via the
 * real ONNX model) and flags cross-shot consistency breaks. */
export async function scoreConsistency(
  shots: ShotInput[],
  options: { session?: ort.InferenceSession; threshold?: number } = {}
): Promise<ConsistencyResult> {
  const session = options.session ?? (await getEmbeddingSession());
  const embedded: { clip: string; embedding: Float32Array }[] = [];
  for (const shot of shots) {
    const embedding = await embedShot(shot.framePaths, session);
    embedded.push({ clip: shot.clip, embedding });
  }
  return flagInconsistentShots(embedded, options.threshold);
}

File: python/src/continuityguard/score/consistency.py
"""
CG02 -- character-consistency scoring.

Computes a per-shot visual embedding and flags cross-shot pairs (claiming
to be the same named character) whose cosine similarity falls below a
threshold derived from this repo's own fixture run (see
src/score/testdata/ and the root CHANGELOG.md's "CG02 fixture
calibration" entry -- never an invented number; the Python port reuses
the exact same threshold since it uses the exact same bundled model file
and the same fixture clips).

IMPORTANT, stated plainly and repeated in the CLI output, the JSON
`reason` field, and the README: this technique is best-validated on
photorealistic/live-action content. Its accuracy on stylized or
anime-adjacent character designs -- the dominant visual style in the
short-drama category this tool targets -- is UNVERIFIED. Treat a flag on
stylized footage as "worth a second look," never as a confirmed defect.

Runtime note: this port uses `onnxruntime` (the Python package), the same
ONNX Runtime project as the TS side's `onnxruntime-node`, loading the
identical bundled `mobilenetv2-7.onnx` file the npm package ships (see
models/NOTICE.md for the model's provenance and the honest limitations of
using a generic ImageNet feature extractor instead of a dedicated face/CLIP
embedding model). Same model file, same runtime family, same preprocessing
-> same embeddings and the same flagged shots as the TypeScript CLI on
identical input.

Ported from src/score/consistency.ts.
"""
from __future__ import annotations

import re
from dataclasses import dataclass, field
from pathlib import Path
from typing import Dict, List, Optional, Sequence

import numpy as np
import onnxruntime as ort

from ..ingest.ffmpeg import FRAME_SIZE, read_raw_frame

_MODEL_PATH = Path(__file__).resolve().parent / "models" / "mobilenetv2-7.onnx"

# ImageNet normalization constants used by the bundled MobileNetV2 model's
# documented preprocessing (see continuityguard/score/models/NOTICE.md).
_IMAGENET_MEAN = (0.485, 0.456, 0.406)
_IMAGENET_STD = (0.229, 0.224, 0.225)

# Cross-shot cosine-similarity threshold below which a shot is flagged as a
# likely character-consistency break. Derived from this repo's own fixture
# run -- see the root CHANGELOG.md "CG02 fixture calibration" entry for the
# exact command and raw numbers this value came from, not an illustrative
# placeholder. Identical to the TS side's CONSISTENCY_SIMILARITY_THRESHOLD.
CONSISTENCY_SIMILARITY_THRESHOLD = 0.88

_cached_session: Optional[ort.InferenceSession] = None


def get_embedding_session() -> ort.InferenceSession:
    """Loads (and caches) the bundled ONNX embedding session. Never fetches
    anything over the network -- the model file ships inside this package."""
    global _cached_session
    if _cached_session is None:
        _cached_session = ort.InferenceSession(
            str(_MODEL_PATH), providers=["CPUExecutionProvider"]
        )
    return _cached_session


def preprocess_frame(raw: bytes) -> np.ndarray:
    """Converts a raw HWC RGB24 frame buffer into the CHW, ImageNet-normalized
    float32 array MobileNetV2 expects, shape (3, FRAME_SIZE, FRAME_SIZE)."""
    size = FRAME_SIZE
    pixels = np.frombuffer(raw, dtype=np.uint8).reshape(size, size, 3).astype(np.float32) / 255.0
    mean = np.array(_IMAGENET_MEAN, dtype=np.float32)
    std = np.array(_IMAGENET_STD, dtype=np.float32)
    normalized = (pixels - mean) / std
    # HWC -> CHW
    return np.transpose(normalized, (2, 0, 1))


def embed_frame(raw: bytes, session: ort.InferenceSession) -> np.ndarray:
    """Runs one frame through the embedding model and returns its raw
    1000-d output vector (see models/NOTICE.md for why this is the raw
    classifier-logit vector rather than a dedicated embedding-layer
    output)."""
    input_array = preprocess_frame(raw)
    tensor = input_array.reshape(1, 3, FRAME_SIZE, FRAME_SIZE)
    input_name = session.get_inputs()[0].name
    output_name = session.get_outputs()[0].name
    results = session.run([output_name], {input_name: tensor})
    return np.asarray(results[0], dtype=np.float32).reshape(-1)


def embed_shot(frame_paths: Sequence[str], session: ort.InferenceSession) -> np.ndarray:
    """Computes one embedding per shot by averaging the embeddings of every
    sampled frame in that shot -- smooths single-frame noise into a more
    stable per-shot signal than picking any single frame would."""
    if not frame_paths:
        raise ValueError("Cannot embed a shot with zero extracted frames")
    total: Optional[np.ndarray] = None
    for frame_path in frame_paths:
        raw = read_raw_frame(frame_path)
        embedding = embed_frame(raw, session)
        total = embedding.copy() if total is None else total + embedding
    assert total is not None
    return total / len(frame_paths)


def cosine_similarity(a: np.ndarray, b: np.ndarray) -> float:
    if a.shape != b.shape:
        raise ValueError("Cannot compare embeddings of different dimensionality")
    norm_a = float(np.linalg.norm(a))
    norm_b = float(np.linalg.norm(b))
    if norm_a == 0 or norm_b == 0:
        return 0.0
    return float(np.dot(a, b) / (norm_a * norm_b))


_CHARACTER_NAME_RE = re.compile(r"^([a-zA-Z][a-zA-Z0-9]*)_")


def parse_character_name(clip_file_name: str) -> Optional[str]:
    """
    Infers the named character a clip belongs to from its filename, using
    a `<character>_<shot-id>.<ext>` convention (e.g. "mei_shot01.mp4" ->
    "mei"). This is a deliberate v0.1 simplification: there is no
    industry-standard character-tagging metadata format across AI
    short-drama pipelines, so ContinuityGuard reads it from the filename
    rather than requiring a separate manifest. Clips that don't match the
    convention return None and are never compared to each other for
    consistency (no named-character claim to check them against).
    """
    without_ext = re.sub(r"\.[^.]+$", "", clip_file_name)
    match = _CHARACTER_NAME_RE.match(without_ext)
    return match.group(1).lower() if match else None


@dataclass(frozen=True)
class ShotInput:
    clip: str
    frame_paths: List[str]


@dataclass(frozen=True)
class ConsistencyFlag:
    clip: str
    character: str
    reference_clip: str
    similarity_score: float
    reason: str


@dataclass(frozen=True)
class ConsistencyResult:
    characters_tracked: int
    flagged_shots: List[ConsistencyFlag] = field(default_factory=list)


def build_consistency_reason(similarity: float) -> str:
    """Builds the honest, unverified-content-labeled reason string for a
    flagged shot -- kept as its own function so every call site (CLI, JSON
    report) uses identical, never-drifting wording."""
    return (
        f"Cross-shot embedding similarity {similarity:.2f} is below the "
        f"{CONSISTENCY_SIMILARITY_THRESHOLD:.2f} threshold. Best-validated on "
        "photorealistic content; accuracy on stylized/anime-adjacent character designs "
        "is unverified -- treat this as a prompt for human review, not a confirmed defect."
    )


def flag_inconsistent_shots(
    shots: Sequence[Dict], threshold: float = CONSISTENCY_SIMILARITY_THRESHOLD
) -> ConsistencyResult:
    """
    Pure grouping + flagging logic, independent of the embedding model --
    given already-computed per-shot embeddings, groups by character and
    flags any shot whose similarity to that character's first-seen
    ("reference") shot falls below threshold. Split out from
    `score_consistency` so this logic has fast, deterministic unit-test
    coverage without depending on real ONNX inference for every test case.

    Each item in `shots` is a dict with keys "clip" (str) and "embedding"
    (np.ndarray).
    """
    by_character: "Dict[str, List[Dict]]" = {}
    for shot in shots:
        character = parse_character_name(shot["clip"])
        if not character:
            continue
        by_character.setdefault(character, []).append(shot)

    flagged_shots: List[ConsistencyFlag] = []
    for character, bucket in by_character.items():
        if len(bucket) < 2:
            continue
        reference = bucket[0]
        for candidate in bucket[1:]:
            similarity = cosine_similarity(reference["embedding"], candidate["embedding"])
            if similarity < threshold:
                flagged_shots.append(
                    ConsistencyFlag(
                        clip=candidate["clip"],
                        character=character,
                        reference_clip=reference["clip"],
                        similarity_score=round(similarity, 4),
                        reason=build_consistency_reason(similarity),
                    )
                )

    return ConsistencyResult(characters_tracked=len(by_character), flagged_shots=flagged_shots)


def score_consistency(
    shots: Sequence[ShotInput],
    session: Optional[ort.InferenceSession] = None,
    threshold: float = CONSISTENCY_SIMILARITY_THRESHOLD,
) -> ConsistencyResult:
    """End-to-end CG02 entry point: computes embeddings for every shot (via
    the real ONNX model) and flags cross-shot consistency breaks."""
    active_session = session or get_embedding_session()
    embedded: List[Dict] = []
    for shot in shots:
        embedding = embed_shot(shot.frame_paths, active_session)
        embedded.append({"clip": shot.clip, "embedding": embedding})
    return flag_inconsistent_shots(embedded, threshold)
